src/Nethermind/Nethermind.Core/Crypto/KeccakCache.cs

```diff
@@ -0,0 +1,226 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Threading;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Core.Crypto;
+
+/// <summary>
+/// This is a minimalistic one-way set associative cache for Keccak values.
+///
+/// It allocates only 8MB of memory to store 64k of entries.
+/// No misaligned reads. Everything is aligned to both cache lines as well as to boundaries so no torn reads.
+/// Requires a single CAS to lock and <see cref="Volatile.Write(ref int,int)"/> to unlock.
+/// On lock failure, it just moves on with execution.
+/// Uses copying on the stack to get the entry, have it copied and release the lock ASAP. This is 128 bytes to copy that quite likely will be the hit.
+/// </summary>
+public static unsafe class KeccakCache
+{
+    /// <summary>
+    /// Count is defined as a +1 over bucket mask. In the future, just change the mask as the main parameter.
+    /// </summary>
+    public const int Count = BucketMask + 1;
+    private const int BucketMask = 0x0000_FFFF;
+
+    private static readonly Entry* Memory;
+
+    static KeccakCache()
+    {
+        const UIntPtr size = Count * Entry.Size;
+
+        // Aligned, so that no torn reads if fields of Entry are properly aligned.
+        Memory = (Entry*)NativeMemory.AlignedAlloc(size, Entry.Size);
+        NativeMemory.Clear(Memory, size);
+    }
+
+    [SkipLocalsInit]
+    public static ValueHash256 Compute(ReadOnlySpan<byte> input)
+    {
+        ComputeTo(input, out ValueHash256 keccak256);
+        return keccak256;
+    }
+
+    [SkipLocalsInit]
+    public static void ComputeTo(ReadOnlySpan<byte> input, out ValueHash256 keccak256)
+    {
+        // Special cases jump forward as unpredicted
+        if (input.Length == 0 || input.Length > Entry.MaxPayloadLength)
+        {
+            goto Uncommon;
+        }
+
+        int hashCode = input.FastHash();
+        uint index = (uint)hashCode & BucketMask;
+
+        Debug.Assert(index < Count);
+
+        ref Entry e = ref Unsafe.Add(ref Unsafe.AsRef<Entry>(Memory), index);
+
+        // Read aligned, volatile, won't be torn, check with computed
+        if (Volatile.Read(ref e.HashCode) == hashCode)
+        {
+            // There's a possibility of a hit, try lock.
+            int lockAndLength = Volatile.Read(ref e.LockAndLength);
+            // Lock by negating length if the length is the same size as input.
+            if (lockAndLength == input.Length && Interlocked.CompareExchange(ref e.LockAndLength, -lockAndLength, lockAndLength) == lockAndLength)
```

Comment:
If you want to go this way, and make the lock less likely in case of a fail, we can squeeze all of the values together. Length is 1 byte long, hash is 2 bytes long, lock can be 1. I'm missing the reasoning in the comment why take this path. Have you found it faster? It's a bit less colliding.